ocaml/xapi/xapi_host.ml

```diff
@@ -1340,10 +1340,98 @@ let get_uncooperative_resident_VMs ~__context ~self = []
+   * belonging to [host] (the term 'host' is overloaded here) *)
+  let open Db_filter_types in
+  let type' = Eq (Field "type", Literal "host") in
+  let host' = Eq (Field "host", Literal (Ref.string_of host)) in
+  let expr = And (type', host') in
+  Db.Certificate.get_records_where ~__context ~expr
+
+let get_ca_cert ~__context ~name : API.ref_Certificate * API.certificate_t =
+  let open Db_filter_types in
+  let type' = Eq (Field "type", Literal "ca") in
+  let name' = Eq (Field "type", Literal name) in
+  let expr = And (type', name') in
+  match Db.Certificate.get_records_where ~__context ~expr with
+  | [x] ->
+      x
+  | xs ->
+      D.error "get_ca_cert: found %i certificates with name='%s'. expected 1"
+        (List.length xs) name ;
+      raise
+        Api_errors.(Server_error (invalid_value, ["certificate:name"; name]))
+
+let add_cert_to_db ~__context ~host ~_type (certificate : X509.Certificate.t) :
+    API.ref_Certificate =
+  let name, _type =
+    (* only user installed CA certs have names *)
+    match _type with `host -> ("", `host) | `ca name -> (name, `ca)
+  in
+  let date_of_ptime time = Date.of_float (Ptime.to_float_s time) in
+  let dates_of_ptimes (a, b) = (date_of_ptime a, date_of_ptime b) in
+  let not_before, not_after =
+    dates_of_ptimes (X509.Certificate.validity certificate)
+  in
+  let fingerprint =
+    X509.Certificate.fingerprint Mirage_crypto.Hash.(`SHA256) certificate
+    |> Certificates.pp_hash
+  in
+  let uuid = Uuid.(to_string (make_uuid ())) in
+  let ref' = Ref.make () in
+  Db.Certificate.create ~__context ~ref:ref' ~uuid ~host ~not_before ~not_after
+    ~fingerprint ~name ~_type ;
+  ref'
+
+let replace_host_certificate ~__context ~host (f : unit -> X509.Certificate.t) :
+    unit =
+  (* NOT atomic *)
+  (* a) create new cert. [f] is assumed to write the cert to the fs
+   * b) add new cert to db
+   * c) remove old certs from fs (maybe of multiple hosts)
+   * d) remove old certs from db
+   * e) restart stunnel *)
+  let old_certs = get_host_certs ~__context ~host in
+  let new_cert = f () in
+  let (_ : API.ref_Certificate) =
+    add_cert_to_db ~__context ~host ~_type:`host new_cert
+  in
+  List.iter
+    (fun (self, record) ->
+      ( match record.API.certificate_type with
+      | `host ->
+          let name = record.API.certificate_uuid in
+          Certificates.(pool_uninstall CA_Certificate) ~__context ~name
+      | `ca ->
+          let name = record.API.certificate_name in
+          Certificates.(host_uninstall CA_Certificate name)
+      ) ;
+      Db.Certificate.destroy ~__context ~self)
+    old_certs ;
+  Xapi_mgmt_iface.reconfigure_stunnel ~__context
+
 let install_ca_certificate ~__context ~host ~name ~cert =
-  Certificates.(host_install CA_Certificate name cert)
+  let certificate = pem_of_string cert in
+  Certificates.(host_install CA_Certificate name cert) ;
+  let (_ : API.ref_Certificate) =
```

Comment:
This call must not modify the database, only pool.install_ca_certificate does when the cert is present in all the hosts' filesystems